File: src/main/java/com/zestarr/pluginportal/utils/HttpUtils.java
package com.zestarr.pluginportal.utils;

import com.fasterxml.jackson.core.JsonParser;
import com.google.gson.JsonObject;
import com.zestarr.pluginportal.PluginPortal;
import com.zestarr.pluginportal.types.LocalPlugin;
import com.zestarr.pluginportal.types.OnlinePlugin;

import javax.net.ssl.HttpsURLConnection;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLConnection;
import java.nio.file.Paths;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class HttpUtils {

    private static String userAgent = "github.com/Zestarr/PluginPortal";

    public static void download(OnlinePlugin onlinePlugin, URL url, File folder) {
        try {

            URLConnection connection = url.openConnection();

            // Set the user agent to "Java" to identify the download as coming from a Java application
            connection.setRequestProperty("User-Agent", userAgent);

            // Get the "Content-Disposition" header field
            String contentDisposition = connection.getHeaderField("Content-Disposition");

            // Extract the default file name from the "Content-Disposition" header field
            String fileName = null;
            if (contentDisposition != null) {
                String[] parts = contentDisposition.split(";");
                for (String part : parts) {
                    if (part.trim().startsWith("filename=")) {
                        fileName = part.substring(part.indexOf('=') + 1);
                        if (fileName.startsWith("\"") && fileName.endsWith("\"")) {
                            fileName = fileName.substring(1, fileName.length() - 1);
                        }
                        break;
                    }
                }
            }

            if (fileName == null) {
                fileName = onlinePlugin.getDefaultFileName();
                if (!fileName.endsWith(".jar")) {
                    fileName = fileName + ".jar";
                }
            }

            // Get an input stream for reading the file
            InputStream inputStream = connection.getInputStream();

            // Create a file output stream to save the file to disk
            FileOutputStream outputStream = new FileOutputStream(new File(folder, fileName));

            // Read the file and save it to disk
            byte[] buffer = new byte[4096];
            int bytesRead = 0;
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, bytesRead);
            }

            // Close the input and output streams
            inputStream.close();
            outputStream.close();

            System.out.println("&7&l[&b&lPPM&7&l] &8&l> Downloaded " + fileName + " to " + folder.getName());
            LocalPlugin localPlugin = new LocalPlugin(onlinePlugin);

            if (PluginPortal.getDeveloperMode()) {
                if (folder == ConfigUtils.getDebugPluginFolder()) {
                    return;
                }
            }

            PluginPortal.getDataManager().getInstalledPlugins().put(localPlugin.getOnlinePlugin().getDisplayName(), localPlugin);


        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void download(OnlinePlugin onlinePlugin, File folder) {
        try {
            download(onlinePlugin, new URL(onlinePlugin.getDownloadLink()), folder);
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
    }

    public static void download(OnlinePlugin onlinePlugin) {
        download(onlinePlugin, ConfigUtils.getPluginFolder());
    }

    public static void download(OnlinePlugin plugin, URL url) {
        download(plugin, url, ConfigUtils.getPluginFolder());
    }

    public static void downloadData(URL url, File file) {
        try {
            // Open a connection to the URL
            URLConnection connection = url.openConnection();

            // Set the user agent to "Java" to identify the download as coming from a Java application
            connection.setRequestProperty("User-Agent", userAgent);

            // Get an input stream for reading the data
            InputStream inputStream = connection.getInputStream();

            // Create a file output stream to write the data to the file
            FileOutputStream outputStream = new FileOutputStream(file);

            // Create a buffered writer to write the data to the file
            BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(outputStream));

            // Read the data and write it to the file
            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
            String line;
            while ((line = reader.readLine()) != null) {
                writer.write(line);
                writer.newLine();
            }

            // Close the input and output streams
            inputStream.close();
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    public static void downloadData(String url, File file) {
        try {
            downloadData(new URL(url), file);
        } catch (MalformedURLException exception) {
            exception.printStackTrace();
        }
    }

    public static void downloadUniversalPlugin(OnlinePlugin plugin) {
        downloadUniversalPlugin(plugin, ConfigUtils.getPluginFolder());
    }

    public static void downloadUniversalPlugin(OnlinePlugin plugin, File folder) {
        String url = plugin.getDownloadLink();
        URL downloadURL = null;

        try {
            downloadURL = new URL(url);
        } catch (MalformedURLException exception) {
            exception.printStackTrace();
        }

        if (downloadURL == null) { return; }

        // Download from SpigotMC
        if (url.contains("https://www.spigotmc.org/resources/")) {
            try {
                download(plugin, new URL("https://api.spiget.org/v2/resources/" + StringUtils.extractNumbers(url) + "/download"), folder);
            } catch (MalformedURLException exception) {
                exception.printStackTrace();
            }

        } else if (url.contains("https://github.com") && url.endsWith(".jar")) {
            download(plugin, downloadURL, folder);
        } else {
            download(plugin, downloadURL, folder);
        }
    }

}

File: src/main/java/com/zestarr/pluginportal/utils/HttpUtils.kt
package com.zestarr.pluginportal.utils

import com.zestarr.pluginportal.types.Plugin
import java.io.*
import java.net.HttpURLConnection
import java.net.URL

// Thanks to SkyClient for some of the code! <3
// Dont dm them for help with this code, dm Zestarr#0001 or report a issue on github.
object HttpUtils {
    //create a simple get request
    operator fun get(url: String?): String? {
        try {
            val obj = URL(url)
            val con = obj.openConnection() as HttpURLConnection
            con.requestMethod = "GET"
            con.setRequestProperty("User-Agent", "github.com/Zestarr/PluginPortalApp")
            val responseCode = con.responseCode
            if (responseCode == 200) {
                val `in` = BufferedReader(InputStreamReader(con.inputStream))
                var inputLine: String?
                val response = StringBuffer()
                while (`in`.readLine().also { inputLine = it } != null) {
                    response.append(inputLine)
                }
                return response.toString()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    fun download(plugin: Plugin, folder: File) {
        println("Downloading " + plugin.downloadLink + " to " + folder.path)
        try {
            val obj = URL(plugin.downloadLink)
            var fileName = ""
            val con = obj.openConnection() as HttpURLConnection
            con.requestMethod = "GET"
            con.setRequestProperty("User-Agent", "github.com/Zestarr/PluginPortalApp")
            con.connectTimeout = 5000
            val responseCode = con.responseCode
            val contentDisposition = con.getHeaderField("Content-Disposition")
            if (contentDisposition != null) {
                val parts = contentDisposition.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                for (part in parts) {
                    if (part.trim { it <= ' ' }.startsWith("filename=")) {
                        fileName = part.substring(part.indexOf('=') + 1)
                        if (fileName.startsWith("\"") && fileName.endsWith("\"")) {
                            fileName = fileName.substring(1, fileName.length - 1)
                        }
                        break
                    }
                }
            }
            if (fileName.isEmpty()) {
                fileName = plugin.displayName + ".jar"
            }
            if (responseCode == 200) {
                val `in` = BufferedInputStream(con.inputStream)
                val fos = FileOutputStream(folder.path + File.separator + fileName)
                val buffer = ByteArray(1024)
                var len: Int
                while (`in`.read(buffer).also { len = it } > 0) {
                    fos.write(buffer, 0, len)
                }
                fos.close()
                `in`.close()
            } else {
                println(responseCode.toString() + " | " + plugin.downloadLink)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun urlEscape(toEncode: String?): String? {
        //escape url
        if (toEncode == null) return null
        val sb = StringBuilder()
        for (character in toEncode.toCharArray()) when (character) {
            '!' -> {
                sb.append("%21")
                continue
            }

            '#' -> {
                sb.append("%23")
                continue
            }

            '$' -> {
                sb.append("%24")
                continue
            }

            '&' -> {
                sb.append("%26")
                continue
            }

            '\'' -> {
                sb.append("%27")
                continue
            }

            '(' -> {
                sb.append("%28")
                continue
            }

            ')' -> {
                sb.append("%29")
                continue
            }

            '*' -> {
                sb.append("%2A")
                continue
            }

            '+' -> {
                sb.append("%2B")
                continue
            }

            ',' -> {
                sb.append("%2C")
                continue
            }

            '/' -> {
                sb.append("%2F")
                continue
            }

            ':' -> {
                sb.append("%3A")
                continue
            }

            ';' -> {
                sb.append("%3B")
                continue
            }

            '=' -> {
                sb.append("%3D")
                continue
            }

            '?' -> {
                sb.append("%3F")
                continue
            }

            '@' -> {
                sb.append("%40")
                continue
            }

            '[' -> {
                sb.append("%5B")
                continue
            }

            ']' -> {
                sb.append("%5D")
                continue
            }

            ' ' -> {
                sb.append("%20")
                continue
            }

            '"' -> {
                sb.append("%22")
                continue
            }

            '%' -> {
                sb.append("%25")
                continue
            }

            '-' -> {
                sb.append("%2D")
                continue
            }

            '<' -> {
                sb.append("%3C")
                continue
            }

            '>' -> {
                sb.append("%3E")
                continue
            }

            '\\' -> {
                sb.append("%5C")
                continue
            }

            '^' -> {
                sb.append("%5E")
                continue
            }

            '_' -> {
                sb.append("%5F")
                continue
            }

            '`' -> {
                sb.append("%60")
                continue
            }

            '{' -> {
                sb.append("%7B")
                continue
            }

            '|' -> {
                sb.append("%7C")
                continue
            }

            '}' -> {
                sb.append("%7D")
                continue
            }

            '~' -> {
                sb.append("%7E")
                continue
            }

            else -> sb.append(character) //if it does not need to be escaped, add the character itself to the StringBuilder
        }
        return sb.toString() //build the string, and return
    }

    fun copyWebsite(websiteURL: String?, filePath: String?) {
        try {
            // Create a URL object for the website
            val website = URL(websiteURL)

            // Open a BufferedReader to read the website's content
            val `in` = BufferedReader(InputStreamReader(website.openStream()))

            // Create a new file at the specified file path
            val file = File(filePath)

            // Check if the file already exists
            if (file.exists()) {
                // If the file exists, delete it
                file.delete()
            }

            // Create a new file
            file.createNewFile()

            // Open a BufferedWriter to write to the file
            val out = BufferedWriter(FileWriter(file))

            // Read each line of the website's content and write it to the file
            var line: String?
            while (`in`.readLine().also { line = it } != null) {
                out.write(line)
                out.newLine()
            }

            // Close the BufferedReader and BufferedWriter
            `in`.close()
            out.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}